### DISCUSSION TITLE
Взаимодействие приложения системы и базы данных

### DISCUSSION THREAD

**Discussion by smplseven** [totumonline/totum-ru-issues-and-discussions#559] [ОШБК!] [answered]
Мы столкнулись с проблемой, связанной с использованием типа NUMERIC в запросах от приложения к базе данных, которая, возможно, негативно влияет на производительность, NUMERIC на наш взгляд препятствует эффективному использованию индексов.

Не могли бы вы подсказать , в чем заключается цель использования NUMERIC  Есть ли конкретные сценарии, где его применение необходимо? Будем признательны за любые ваши рекомендации или пояснения.

Вот пример для  сравнения: 

Разница между двумя запросами заключается в типе данных и в том, как это влияет на использование индекса:

 
1. **Текущий запрос:**
   ```sql
   explain select id from vyplaty where ((id)::NUMERIC = 2001403 ) limit 1;
   ```
   В этом запросе `id` явно приводится к типу `NUMERIC`, что меняет способ выполнения запроса. Приведение типов не позволяет использовать индекс напрямую, так как происходит преобразование данных. План запроса показывает:
   ```
   Limit  (cost=0.43..6.07 rows=1 width=4)
     ->  Index Only Scan using vyplaty_pkey on vyplaty  (cost=0.43..73047.87 rows=12950 width=4)
           Filter: ((id)::numeric = '2001403'::numeric)
   ```
   В этом случае, хотя всё ещё используется `Index Only Scan`, фильтрация выполняется после сканирования индекса, что увеличивает стоимость и время выполнения запроса. Это происходит из-за необходимости преобразования каждого значения `id` к типу `NUMERIC` для сравнения.

2.**Возможный запрос:**
   ```sql
   explain select id from vyplaty where ((id) = 2001403 ) limit 1;
   ```
   В этом запросе `id` используется как есть, без явного приведения типа. Предполагая, что `id` в таблице `vyplaty` имеет тип `integer` (или любой другой индексируемый тип, поддерживающий прямое сопоставление), база данных может использовать индекс для быстрого поиска. План запроса показывает:
   ```
   Limit  (cost=0.43..2.65 rows=1 width=4)
     ->  Index Only Scan using vyplaty_pkey on vyplaty  (cost=0.43..2.65 rows=1 width=4)
           Index Cond: (id = 2001403)
   ```
   Здесь видно, что база данных использует `Index Only Scan`, что говорит о прямом использовании индекса для поиска.

**Comment by poluninatm** [comment]
Добрый день. Спасибо за исследование. Приведение типа для поля id используется в рамках общей унификации запросов к числовым полям Тотума. Для ид следует сделать исключение, но когда это будет реализовано я не знаю.

**Reply by ttm-alexeypolunin** [reply]
~ в августе-сентябре

**Reply by ttm-alexeypolunin** [reply]
А я то думал, какого черта он тормозит на id :)

**Reply by smplseven** [reply]
Большое спасибо за ответ!

**Reply by ttm-alexeypolunin** [reply, marked as answer]
Вроде как исправлено в V5
